Implement getUniqueId helper function

livetocode opened the pull request:
which can generate different kinds of unique IDs from a JWT, based on your requirements.
The purpose is to normalize how we build such IDs in our projects.
Also, the introduction of the URN format would help diagnosis or audits since it contains more information.

livetocode committed:
Implement getUniqueId helper function
which can generate different kinds of unique IDs
from a JWT, based on your requirements.

Signed-off-by: Morgan Martinet <morgan.martinet@montreal.ca>

steph-lebl commented:
Here are my two cents: **We should avoid adding new user identifier formats and promote using only HumanReadable user identifier by convention.** 

1) HumanReadable user identifiers are unperfect, but they are largely used by existing systems and for integration between them. We will have to support them anyway and adding new formats will only create more work and risk of errors. Moreover, HumanReadableURN does not really provide more information as it is possible to infer most of this information from the HumanReadable format.

2) Opaque IDs are not stable and would be lost if we change the authentication provider. This is unacceptable, because if will create large integration problem between systems when this will occurs (it will occurs eventually, even if we are very much in love with MS right now and the wedding was very expensive).

Lastly, CodeU or emails can be reassigned once the person leaves. We must change this as soon as possible, because this bad practice creates way more harm than good. **What can we do about it?**

livetocode committed:
remove some validation errors to avoid issues
if we ever add new realms or environments.
Otherwise we would have to patch the lib and
redeploy all apps that depend on it before
adding new values.

Signed-off-by: Morgan Martinet <morgan.martinet@montreal.ca>

livetocode committed:
do not use oid for client IDs

Signed-off-by: Morgan Martinet <morgan.martinet@montreal.ca>

livetocode committed:
minor refactoring after code review

Signed-off-by: Morgan Martinet <morgan.martinet@montreal.ca>
